--- a/net/minecraft/command/impl/SetBlockCommand.java
+++ b/net/minecraft/command/impl/SetBlockCommand.java
@@ -1,5 +1,7 @@
 package net.minecraft.command.impl;
 
+import carpet.helpers.CommandPermissionHelper;
+import carpet.settings.CarpetSettings;
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
@@ -24,11 +26,10 @@
 
     public static void register(CommandDispatcher<CommandSource> dispatcher)
     {
-        dispatcher.register(Commands.literal("setblock").requires((p_198688_0_) ->
+        dispatcher.register(Commands.literal("setblock").requires(
+            CommandPermissionHelper::canCheat
+        ).then(Commands.argument("pos", BlockPosArgument.blockPos()).then(Commands.argument("block", BlockStateArgument.blockState()).executes((p_198682_0_) ->
         {
-            return p_198688_0_.hasPermissionLevel(2);
-        }).then(Commands.argument("pos", BlockPosArgument.blockPos()).then(Commands.argument("block", BlockStateArgument.blockState()).executes((p_198682_0_) ->
-        {
             return setBlock(p_198682_0_.getSource(), BlockPosArgument.getLoadedBlockPos(p_198682_0_, "pos"), BlockStateArgument.getBlockState(p_198682_0_, "block"), SetBlockCommand.Mode.REPLACE, (Predicate<BlockWorldState>)null);
         }).then(Commands.literal("destroy").executes((p_198685_0_) ->
         {
@@ -73,13 +74,18 @@
                 flag = true;
             }
 
-            if (flag && !state.place(worldserver, pos, 2))
+            if (flag && !state.place(worldserver, pos, 2 | (CarpetSettings.fillUpdates?0:1024))) // [CM]
             {
                 throw FAILED_EXCEPTION.create();
             }
             else
             {
+                // CM
+                if (CarpetSettings.fillUpdates)
+                {
                 worldserver.notifyNeighbors(pos, state.getState().getBlock());
+                }
+                // CM end
                 source.sendFeedback(new TextComponentTranslation("commands.setblock.success", pos.getX(), pos.getY(), pos.getZ()), true);
                 return 1;
             }
