--- a/net/minecraft/command/impl/TeleportCommand.java
+++ b/net/minecraft/command/impl/TeleportCommand.java
@@ -1,5 +1,6 @@
 package net.minecraft.command.impl;
 
+import carpet.helpers.CommandPermissionHelper;
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.tree.LiteralCommandNode;
@@ -30,11 +31,10 @@
 {
     public static void register(CommandDispatcher<CommandSource> dispatcher)
     {
-        LiteralCommandNode<CommandSource> literalcommandnode = dispatcher.register(Commands.literal("teleport").requires((p_198816_0_) ->
+        LiteralCommandNode<CommandSource> literalcommandnode = dispatcher.register(Commands.literal("teleport").requires(
+            CommandPermissionHelper::canCheat  // TISCM opPlayerNoCheat
+        ).then(Commands.argument("targets", EntityArgument.entities()).then(Commands.argument("location", Vec3Argument.vec3()).executes((p_198807_0_) ->
         {
-            return p_198816_0_.hasPermissionLevel(2);
-        }).then(Commands.argument("targets", EntityArgument.entities()).then(Commands.argument("location", Vec3Argument.vec3()).executes((p_198807_0_) ->
-        {
             return teleportToPos(p_198807_0_.getSource(), EntityArgument.getEntities(p_198807_0_, "targets"), p_198807_0_.getSource().getWorld(), Vec3Argument.getLocation(p_198807_0_, "location"), (ILocationArgument)null, (TeleportCommand.Facing)null);
         }).then(Commands.argument("rotation", RotationArgument.rotation()).executes((p_198811_0_) ->
         {
@@ -58,10 +58,9 @@
         {
             return teleportToEntity(p_200562_0_.getSource(), Collections.singleton(p_200562_0_.getSource().assertIsEntity()), EntityArgument.getEntity(p_200562_0_, "destination"));
         })));
-        dispatcher.register(Commands.literal("tp").requires((p_200556_0_) ->
-        {
-            return p_200556_0_.hasPermissionLevel(2);
-        }).redirect(literalcommandnode));
+        dispatcher.register(Commands.literal("tp").requires(
+            CommandPermissionHelper::canCheat  // TISCM opPlayerNoCheat
+        ).redirect(literalcommandnode));
     }
 
     private static int teleportToEntity(CommandSource source, Collection <? extends Entity > targets, Entity destination)
