--- a/net/minecraft/entity/projectile/EntitySnowball.java
+++ b/net/minecraft/entity/projectile/EntitySnowball.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.projectile;
 
+import carpet.logging.LoggerRegistry;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.EntityType;
 import net.minecraft.entity.monster.EntityBlaze;
@@ -56,7 +57,28 @@
         if (!this.world.isRemote)
         {
             this.world.setEntityState(this, (byte)3);
+
+            //TISCM add collide at logger
+            if (LoggerRegistry.__projectiles && this.logHelper != null)
+            {
+                if (result.type == RayTraceResult.Type.BLOCK) {
+                    this.logHelper.onCollide(result.hitVec.x, result.hitVec.y, result.hitVec.z, "Block", this.world);
+                }
+                else {
+                    this.logHelper.onCollide(result.hitVec.x, result.hitVec.y, result.hitVec.z, "Entity", this.world);
+                }
+            }
+            //END TISCM
+
             this.remove();
         }
     }
+
+    //TISCM make sure visualize projectile logger snowballs cannt get moved
+    public boolean isImmuneToExplosions()
+    {
+        if (this.getTags().contains("TISCM_VISPROJ_LOGGER")) return true;
+        else return false;
+    }
+    //end TISCM
 }
