--- a/net/minecraft/world/chunk/storage/AnvilChunkLoader.java
+++ b/net/minecraft/world/chunk/storage/AnvilChunkLoader.java
@@ -1,5 +1,7 @@
 package net.minecraft.world.chunk.storage;
 
+import carpet.settings.CarpetSettings;
+import carpet.utils.LightingHooks;
 import com.google.common.collect.Maps;
 import com.mojang.datafixers.DataFixTypes;
 import com.mojang.datafixers.DataFixer;
@@ -483,6 +485,9 @@
         ChunkSection[] achunksection = chunkIn.getSections();
         NBTTagList nbttaglist = this.writeChunkSectionsToNBT(worldIn, achunksection);
         compound.put("Sections", nbttaglist);
+        // [TISCM] Newlight
+        if (CarpetSettings.newLight)
+            LightingHooks.writeLightData(chunkIn, compound);
         Biome[] abiome = chunkIn.getBiomes();
         int[] aint = new int[abiome.length];
 
@@ -616,6 +621,9 @@
         chunk.setStatus(compound.getString("Status"));
         NBTTagList nbttaglist = compound.getList("Sections", 10);
         chunk.setSections(this.readSectionsFromNBT(worldIn, nbttaglist));
+        // [TISCM] Newlight
+        if (CarpetSettings.newLight)
+            LightingHooks.readLightData(chunk, compound);
         NBTTagCompound nbttagcompound = compound.getCompound("Heightmaps");
 
         for (Heightmap.Type heightmap$type : Heightmap.Type.values())
