--- a/net/minecraft/entity/EntityTrackerEntry.java
+++ b/net/minecraft/entity/EntityTrackerEntry.java
@@ -5,6 +5,8 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
+
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import net.minecraft.block.Block;
 import net.minecraft.entity.ai.attributes.AttributeMap;
 import net.minecraft.entity.ai.attributes.IAttributeInstance;
@@ -94,7 +96,11 @@
     private boolean ridingEntity;
     private boolean onGround;
     public boolean playerEntitiesUpdated;
-    private final Set<EntityPlayerMP> trackingPlayers = Sets.newHashSet();
+    /**
+     * Lithium change:
+     * Uses less memory, and will cache the returned iterator
+     */
+    private final Set<EntityPlayerMP> trackingPlayers = new ObjectOpenHashSet<EntityPlayerMP>();
 
     public EntityTrackerEntry(Entity entityIn, int rangeIn, int maxRangeIn, int updateInterval, boolean sendVelocityUpdatesIn)
     {
