--- a/net/minecraft/client/multiplayer/ChunkProviderClient.java
+++ b/net/minecraft/client/multiplayer/ChunkProviderClient.java
@@ -1,5 +1,7 @@
 package net.minecraft.client.multiplayer;
 
+import carpet.settings.CarpetSettings;
+import carpet.utils.LightingHooks;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMaps;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
@@ -69,6 +71,9 @@
                 return new Chunk(this.world, x, z, new Biome[256]);
             });
             chunk.read(buf, p_212474_4_, p_212474_5_);
+            // [TISCM] Newlight
+            if (CarpetSettings.newLight)
+                LightingHooks.onLoadClient(chunk.getWorld(), chunk);
             chunk.setLoaded(true);
             return chunk;
         }
