--- a/net/minecraft/entity/item/EntityXPOrb.java
+++ b/net/minecraft/entity/item/EntityXPOrb.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.item;
 
+import carpet.settings.CarpetSettings;
 import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
@@ -17,11 +18,14 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
+import carpet.helpers.XPcombine;
+
 public class EntityXPOrb extends Entity
 {
     public int xpColor;
     public int xpOrbAge;
     public int delayBeforeCanPickup;
+    public int delayBeforeCombine;
     private int xpOrbHealth = 5;
     private int xpValue;
     private EntityPlayer closestPlayer;
@@ -37,8 +41,26 @@
         this.motionY = (double)((float)(Math.random() * 0.2D) * 2.0F);
         this.motionZ = (double)((float)(Math.random() * (double)0.2F - (double)0.1F) * 2.0F);
         this.xpValue = expValue;
+
+        // add combine delay [CM]
+        this.delayBeforeCombine = 50;
     }
 
+    public EntityXPOrb(World worldIn, int expValue, EntityXPOrb old)
+    {
+        super(EntityType.EXPERIENCE_ORB, worldIn);
+        this.setSize(0.5F, 0.5F);
+        this.setPosition(old.posX, old.posY, old.posZ);
+        this.rotationYaw = old.rotationYaw;
+        this.motionX = old.motionX;
+        this.motionY = old.motionY;
+        this.motionZ = old.motionZ;
+        this.xpValue = expValue;
+
+        // add combine delay [CM]
+        this.delayBeforeCombine = 50;
+    }
+
     public EntityXPOrb(World worldIn)
     {
         super(EntityType.EXPERIENCE_ORB, worldIn);
@@ -103,28 +125,28 @@
         }
 
         this.pushOutOfBlocks(this.posX, (this.getBoundingBox().minY + this.getBoundingBox().maxY) / 2.0D, this.posZ);
-        double d0 = 8.0D;
+        double d0 = CarpetSettings.xpTrackingDistance; // TISCM xpTrackingDistance
 
         if (this.xpTargetColor < this.xpColor - 20 + this.getEntityId() % 100)
         {
-            if (this.closestPlayer == null || this.closestPlayer.getDistanceSq(this) > 64.0D)
+            if (this.closestPlayer == null || this.closestPlayer.getDistanceSq(this) > d0 * d0)
             {
-                this.closestPlayer = this.world.getClosestPlayerToEntity(this, 8.0D);
+                this.closestPlayer = this.world.getClosestPlayerToEntity(this, d0);
             }
 
             this.xpTargetColor = this.xpColor;
         }
 
-        if (this.closestPlayer != null && this.closestPlayer.isSpectator())
+        if ((this.closestPlayer != null && this.closestPlayer.isSpectator()) || d0 == 0.0D) // TISCM xpTrackingDistance
         {
             this.closestPlayer = null;
         }
 
         if (this.closestPlayer != null)
         {
-            double d1 = (this.closestPlayer.posX - this.posX) / 8.0D;
-            double d2 = (this.closestPlayer.posY + (double)this.closestPlayer.getEyeHeight() / 2.0D - this.posY) / 8.0D;
-            double d3 = (this.closestPlayer.posZ - this.posZ) / 8.0D;
+            double d1 = (this.closestPlayer.posX - this.posX) / d0;
+            double d2 = (this.closestPlayer.posY + (double)this.closestPlayer.getEyeHeight() / 2.0D - this.posY) / d0;
+            double d3 = (this.closestPlayer.posZ - this.posZ) / d0;
             double d4 = Math.sqrt(d1 * d1 + d2 * d2 + d3 * d3);
             double d5 = 1.0D - d4;
 
@@ -138,6 +160,21 @@
         }
 
         this.move(MoverType.SELF, this.motionX, this.motionY, this.motionZ);
+
+        //Adding code here for combining XP orbs. [CM]
+        if (CarpetSettings.combineXPOrbs)
+        {
+            if (this.delayBeforeCombine > 0)
+            {
+                --this.delayBeforeCombine;
+            }
+
+            if (this.delayBeforeCombine == 0)
+            {
+                XPcombine.searchForOtherXPNearbyCarpet(this);
+            }
+        }
+
         float f = 0.98F;
 
         if (this.onGround)
@@ -218,7 +255,8 @@
     {
         if (!this.world.isRemote)
         {
-            if (this.delayBeforeCanPickup == 0 && entityIn.xpCooldown == 0)
+            // [CM] added ORed condition
+            if (this.delayBeforeCanPickup == 0 && (entityIn.xpCooldown == 0 || CarpetSettings.xpNoCooldown))
             {
                 entityIn.xpCooldown = 2;
                 entityIn.onItemPickup(this, 1);
@@ -251,6 +289,10 @@
         return xp * 2;
     }
 
+    public void setXpValue(int xpValue)
+    {
+        this.xpValue = xpValue;
+    }
     public int getXpValue()
     {
         return this.xpValue;
