--- a/net/minecraft/entity/item/EntityEnderPearl.java
+++ b/net/minecraft/entity/item/EntityEnderPearl.java
@@ -1,6 +1,9 @@
 package net.minecraft.entity.item;
 
 import javax.annotation.Nullable;
+
+import carpet.logging.LoggerRegistry;
+import carpet.patches.EntityPlayerMPFake;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
@@ -92,7 +95,8 @@
             {
                 EntityPlayerMP entityplayermp = (EntityPlayerMP)entitylivingbase;
 
-                if (entityplayermp.connection.getNetworkManager().isChannelOpen() && entityplayermp.world == this.world && !entityplayermp.isPlayerSleeping())
+                //CM added condition for fake players to throw enderpearl
+                if ((entityplayermp.connection.getNetworkManager().isChannelOpen() || entityplayermp instanceof EntityPlayerMPFake) && entityplayermp.world == this.world && !entityplayermp.isPlayerSleeping())
                 {
                     if (this.rand.nextFloat() < 0.05F && this.world.getGameRules().getBoolean("doMobSpawning"))
                     {
@@ -118,6 +122,15 @@
                 entitylivingbase.fallDistance = 0.0F;
             }
 
+            if (LoggerRegistry.__projectiles && this.logHelper != null)
+            {
+                if (result.type == RayTraceResult.Type.BLOCK) {
+                    this.logHelper.onCollide(result.hitVec.x, result.hitVec.y, result.hitVec.z, "Block");
+            }
+                else {
+                    this.logHelper.onCollide(result.hitVec.x, result.hitVec.y, result.hitVec.z, "Entity");
+                }
+            }
             this.remove();
         }
     }
