--- a/net/minecraft/world/Teleporter.java
+++ b/net/minecraft/world/Teleporter.java
@@ -1,5 +1,8 @@
 package net.minecraft.world;
 
+import carpet.utils.portalsearcher.*;
+import carpet.utils.portalsearcher.PortalSearcherSuperCache;
+import carpet.settings.CarpetSettings;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.objects.ObjectIterator;
@@ -10,6 +13,7 @@
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.Blocks;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
@@ -22,6 +26,8 @@
     private final WorldServer world;
     private final Random random;
     private final Long2ObjectMap<Teleporter.PortalPosition> destinationCoordinateCache = new Long2ObjectOpenHashMap<>(4096);
+    //CM var portalCaching
+    private final Long2ObjectMap<Teleporter.PortalPosition> destinationHistoryCache = new Long2ObjectOpenHashMap<>(4096); //carpet mod
 
     public Teleporter(WorldServer worldIn)
     {
@@ -76,6 +82,7 @@
         int j = MathHelper.floor(entityIn.posX);
         int k = MathHelper.floor(entityIn.posZ);
         boolean flag = true;
+        boolean flag_cm = true; // CM portalCaching
         BlockPos blockpos = BlockPos.ORIGIN;
         long l = ChunkPos.asLong(j, k);
 
@@ -87,38 +94,46 @@
             teleporter$portalposition.lastUpdateTime = this.world.getGameTime();
             flag = false;
         }
-        else
+        else if (CarpetSettings.portalCaching && this.destinationHistoryCache.containsKey(l)) // potential best candidate for linkage.
         {
-            BlockPos blockpos3 = new BlockPos(entityIn);
-
-            for (int i1 = -128; i1 <= 128; ++i1)
+            Teleporter.PortalPosition teleporter$portalposition_cm = (Teleporter.PortalPosition)this.destinationHistoryCache.get(l);
+            if (this.world.getBlockState(teleporter$portalposition_cm).getBlock() == Blocks.NETHER_PORTAL)
+            //just to verify nobody is cheating the system with update suppression
             {
-                BlockPos blockpos2;
+                d0 = 0.0D;
+                blockpos = teleporter$portalposition_cm;
+                flag_cm = false;
+            }
+        }
 
-                for (int j1 = -128; j1 <= 128; ++j1)
-                {
-                    for (BlockPos blockpos1 = blockpos3.add(i1, this.world.getActualHeight() - 1 - blockpos3.getY(), j1); blockpos1.getY() >= 0; blockpos1 = blockpos2)
-                    {
-                        blockpos2 = blockpos1.down();
+        //else portalCaching
+        if (d0 < 0.0D)
+        {
 
-                        if (this.world.getBlockState(blockpos1).getBlock() == BLOCK_NETHER_PORTAL)
-                        {
-                            for (blockpos2 = blockpos1.down(); this.world.getBlockState(blockpos2).getBlock() == BLOCK_NETHER_PORTAL; blockpos2 = blockpos2.down())
-                            {
-                                blockpos1 = blockpos2;
-                            }
+            // TISCM betterPortalSearcher
+            PortalSearcherAbstract searcher;
+            switch (CarpetSettings.betterPortalSearcher) {
+            case BOX: // similar to spiral, but slightly faster
+                searcher = new PortalSearcherBox(world);
+                break;
+            case SPIRAL:
+                searcher = new PortalSearcherSpiral(world);
+                break;
+            case SUPER_CACHE: // slightly slower than vanilla for the first time, but can be super fast later on
+                searcher = new PortalSearcherSuperCache(world);
+                break;
 
-                            double d1 = blockpos1.distanceSq(blockpos3);
-
-                            if (d0 < 0.0D || d1 < d0)
-                            {
-                                d0 = d1;
-                                blockpos = blockpos1;
-                            }
-                        }
-                    }
-                }
+            case VANILLA:
+            default: // it should be the same as vanilla
+                searcher = new PortalSearcherVanilla(world);
+                break;
             }
+            PortalSearchResult result = searcher.search(new BlockPos(entityIn));
+            d0 = result.getDistanceSq();
+            blockpos = result.getResult();
+//            LogManager.getLogger().info(result.toString()); // only for debug
+
+            // end TISCM betterPortalSearcher
         }
 
         if (d0 >= 0.0D)
@@ -128,6 +143,15 @@
                 this.destinationCoordinateCache.put(l, new Teleporter.PortalPosition(blockpos, this.world.getGameTime()));
             }
 
+            /* carpet mod */
+            if (CarpetSettings.portalCaching)
+            {
+                //its timeless
+                this.destinationHistoryCache.put(l, new Teleporter.PortalPosition(blockpos, 0L));
+                this.destinationCoordinateCache.put(l, new Teleporter.PortalPosition(blockpos, this.world.getGameTime()));
+            }
+            /* carpet mod end */
+
             double d5 = (double)blockpos.getX() + 0.5D;
             double d7 = (double)blockpos.getZ() + 0.5D;
             BlockPattern.PatternHelper blockpattern$patternhelper = BLOCK_NETHER_PORTAL.createPatternHelper(this.world, blockpos);
@@ -139,14 +163,35 @@
             {
                 ++d2;
             }
-
+    
+            // [CM] portalSuffocationFix - removed offset calculation outside of the if statement
+            double offset = (1.0D - entityIn.getLastPortalVec().x) * (double)blockpattern$patternhelper.getWidth() * (double)blockpattern$patternhelper.getForwards().rotateY().getAxisDirection().getOffset();
+            if (CarpetSettings.portalSuffocationFix)
+            {
+                double entity_corrected_radius = 1.02*(double)entityIn.width/2;
+                if (entity_corrected_radius >= (double)blockpattern$patternhelper.getWidth()-entity_corrected_radius)
+                {
+                    //entity is wider than portal, so will suffocate anyways, so place it directly in the middle
+                    entity_corrected_radius = (double)blockpattern$patternhelper.getWidth()/2-0.001;
+                }
+        
+                if (offset >= 0)
+                {
+                    offset = MathHelper.clamp(offset, entity_corrected_radius, (double)blockpattern$patternhelper.getWidth()-entity_corrected_radius);
+                }
+                else
+                {
+                    offset = MathHelper.clamp(offset, -(double)blockpattern$patternhelper.getWidth()+entity_corrected_radius, -entity_corrected_radius);
+                }
+            }
+            
             if (blockpattern$patternhelper.getForwards().getAxis() == EnumFacing.Axis.X)
             {
-                d7 = d2 + (1.0D - entityIn.getLastPortalVec().x) * (double)blockpattern$patternhelper.getWidth() * (double)blockpattern$patternhelper.getForwards().rotateY().getAxisDirection().getOffset();
+                d7 = d2 + offset;
             }
             else
             {
-                d5 = d2 + (1.0D - entityIn.getLastPortalVec().x) * (double)blockpattern$patternhelper.getWidth() * (double)blockpattern$patternhelper.getForwards().rotateY().getAxisDirection().getOffset();
+                d5 = d2 + offset;
             }
 
             float f = 0.0F;
@@ -394,6 +439,13 @@
             }
         }
 
+        /* carpet mod */
+        if (CarpetSettings.portalCaching)
+        {
+            clearHistoryCache();
+        }
+        /* carpet mod */
+
         return true;
     }
 
@@ -413,9 +465,37 @@
                     objectiterator.remove();
                 }
             }
+            /* carpet mod */
+            //failsafe - arbitrary, but will never happen in normal circumstances,
+            //but who knows these freekin players.
+            if (CarpetSettings.portalCaching && this.destinationHistoryCache.size()> 65000 )
+            {
+                removeAllCachedEntries();
+            }
+
+            // TISCM betterPortalSearcher
+            if (CarpetSettings.betterPortalSearcher != CarpetSettings.EnumPortalSearcher.SUPER_CACHE) {
+                (this.world.getDimension().isNether() ? SuperCacheHandler.getHandlerNether() : SuperCacheHandler.getHandlerOverworld()).clear();
+            }
+            // end TISCM betterPortalSearcher
+            /* carpet mod end */
         }
     }
 
+    //carpet mod portalCaching
+    public void clearHistoryCache()
+    {
+        MinecraftServer minecraftserver = this.world.getServer();
+        minecraftserver.getWorlds().forEach( (w) -> w.getDefaultTeleporter().removeAllCachedEntries());
+    }
+
+    //carpet mod portalCaching
+    public void removeAllCachedEntries()
+    {
+        this.destinationHistoryCache.clear();
+    }
+
+
     public class PortalPosition extends BlockPos
     {
         public long lastUpdateTime;
