--- a/net/minecraft/entity/boss/dragon/phase/PhaseSittingFlaming.java
+++ b/net/minecraft/entity/boss/dragon/phase/PhaseSittingFlaming.java
@@ -1,5 +1,6 @@
 package net.minecraft.entity.boss.dragon.phase;
 
+import carpet.settings.CarpetSettings;
 import net.minecraft.entity.EntityAreaEffectCloud;
 import net.minecraft.entity.boss.EntityDragon;
 import net.minecraft.init.MobEffects;
@@ -75,6 +76,15 @@
             while (this.dragon.world.isAirBlock(blockpos$mutableblockpos))
             {
                 --d2;
+
+                //TISCM fix for dragon crash
+                if (CarpetSettings.dragonCrashFix) {
+                    if (d2 < 0) {
+                        d2 = this.dragon.dragonPartHead.posX - 6;
+                        break;
+                    }
+                }
+
                 blockpos$mutableblockpos.setPos(MathHelper.floor(d0), MathHelper.floor(d2), MathHelper.floor(d1));
             }
 
