--- a/net/minecraft/block/BlockTripWireHook.java
+++ b/net/minecraft/block/BlockTripWireHook.java
@@ -166,6 +166,7 @@
         {
             BlockPos blockpos1 = pos.offset(enumfacing, i);
             EnumFacing enumfacing1 = enumfacing.getOpposite();
+            boolean oldPoweredState = worldIn.getBlockState(blockpos1).get(POWERED);
             worldIn.setBlockState(blockpos1, iblockstate1.with(FACING, enumfacing1), 3);
             this.notifyNeighbors(worldIn, blockpos1, enumfacing1);
             this.playSound(worldIn, blockpos1, flag2, flag3, flag, flag1);
