--- a/net/minecraft/block/BlockRedstoneComparator.java
+++ b/net/minecraft/block/BlockRedstoneComparator.java
@@ -3,6 +3,8 @@
 import java.util.List;
 import java.util.Random;
 import javax.annotation.Nullable;
+
+import carpet.logging.logHelpers.MicroTickLogHelper;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.item.EntityItemFrame;
 import net.minecraft.entity.player.EntityPlayer;
@@ -144,6 +146,7 @@
             {
                 TickPriority tickpriority = this.isFacingTowardsRepeater(worldIn, pos, state) ? TickPriority.HIGH : TickPriority.NORMAL;
                 worldIn.getPendingBlockTicks().scheduleTick(pos, this, 2, tickpriority);
+                MicroTickLogHelper.onComponentAddToTileTickList(worldIn, pos, 2, tickpriority); // TISCM Micro Tick logger
             }
         }
     }
@@ -169,10 +172,12 @@
             if (flag && !flag1)
             {
                 worldIn.setBlockState(pos, state.with(POWERED, Boolean.valueOf(false)), 2);
+                MicroTickLogHelper.onComponentPowered(worldIn, pos, false); // TISCM Micro Tick logger
             }
             else if (!flag && flag1)
             {
                 worldIn.setBlockState(pos, state.with(POWERED, Boolean.valueOf(true)), 2);
+                MicroTickLogHelper.onComponentPowered(worldIn, pos, true); // TISCM Micro Tick logger
             }
 
             this.notifyNeighbors(worldIn, pos, state);
