--- a/net/minecraft/block/Block.java
+++ b/net/minecraft/block/Block.java
@@ -175,7 +175,11 @@
     @Deprecated
     public MaterialColor getMaterialColor(IBlockState state, IBlockReader worldIn, BlockPos pos)
     {
-        return this.materialColor;
+    	if (carpet.settings.CarpetSettings.stainedGlassNoMapRendering && state.getMaterial() == state.getMaterial().GLASS)
+    	{
+    		return this.materialColor.AIR;
+    	}
+    	return this.materialColor;
     }
 
     @Deprecated
@@ -1764,13 +1768,28 @@
             private final IBlockState adjacentState;
             private final EnumFacing side;
 
+            // Lithium
+            private int hash;
+
             public RenderSideCacheKey(IBlockState state, IBlockState adjacentState, EnumFacing side)
             {
                 this.state = state;
                 this.adjacentState = adjacentState;
                 this.side = side;
+                this.genHash();
             }
 
+            /**
+             * Initialize the object's hashcode and cache it
+             */
+            private void genHash() {
+                int hash = this.state.hashCode();
+                hash = 31 * hash + this.adjacentState.hashCode();
+                hash = 31 * hash + this.side.hashCode();
+
+                this.hash = hash;
+            }
+
             public boolean equals(Object p_equals_1_)
             {
                 if (this == p_equals_1_)
@@ -1790,7 +1809,8 @@
 
             public int hashCode()
             {
-                return Objects.hash(this.state, this.adjacentState, this.side);
+                // return Objects.hash(this.state, this.adjacentState, this.side);
+                return this.hash; //Lithium
             }
         }
 }
