--- a/net/minecraft/block/BlockPistonBase.java
+++ b/net/minecraft/block/BlockPistonBase.java
@@ -1,5 +1,6 @@
 package net.minecraft.block;
 
+import carpet.settings.CarpetSettings;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import java.util.List;
@@ -15,10 +16,12 @@
 import net.minecraft.item.BlockItemUseContext;
 import net.minecraft.item.ItemStack;
 import net.minecraft.pathfinding.PathType;
+import net.minecraft.server.management.PlayerList;
 import net.minecraft.state.BooleanProperty;
 import net.minecraft.state.StateContainer;
 import net.minecraft.state.properties.BlockStateProperties;
 import net.minecraft.state.properties.PistonType;
+import net.minecraft.stats.StatList;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.tileentity.TileEntityPiston;
 import net.minecraft.util.EnumFacing;
@@ -158,7 +161,7 @@
 
     private boolean shouldBeExtended(World worldIn, BlockPos pos, EnumFacing facing)
     {
-        for (EnumFacing enumfacing : EnumFacing.values())
+        for (EnumFacing enumfacing : EnumFacing.VALUES)
         {
             if (enumfacing != facing && worldIn.isSidePowered(pos.offset(enumfacing), enumfacing))
             {
@@ -172,9 +175,16 @@
         }
         else
         {
+            // [CM] QuasiConnectivity Tweak For Pistons
+            if (!CarpetSettings.quasiConnectivity)
+            {
+                return false;
+            }
+            // [CM] End
+
             BlockPos blockpos = pos.up();
 
-            for (EnumFacing enumfacing1 : EnumFacing.values())
+            for (EnumFacing enumfacing1 : EnumFacing.VALUES)
             {
                 if (enumfacing1 != EnumFacing.DOWN && worldIn.isSidePowered(blockpos.offset(enumfacing1), enumfacing1))
                 {
@@ -186,6 +196,15 @@
         }
     }
 
+    // TISCM break_bedrock stat
+    private void tryAddNearestPlayerStats(World worldIn, BlockPos pos)
+    {
+        if (worldIn.getBlockState(pos).getBlock() == Blocks.BEDROCK)
+        {
+            worldIn.getServer().getPlayerList().addStatsToNearestPlayersInDimensionNear(worldIn.getDimension().getType(), pos, 10, StatList.BREAK_BEDROCK, 1);
+        }
+    }
+
     public boolean eventReceived(IBlockState state, World worldIn, BlockPos pos, int id, int param)
     {
         EnumFacing enumfacing = state.get(FACING);
@@ -255,6 +274,9 @@
                 {
                     if (id != 1 || iblockstate.isAir() || !canPush(iblockstate, worldIn, blockpos, enumfacing.getOpposite(), false, enumfacing) || iblockstate.getPushReaction() != EnumPushReaction.NORMAL && block != Blocks.PISTON && block != Blocks.STICKY_PISTON)
                     {
+                        // TISCM break_bedrock stat
+                        tryAddNearestPlayerStats(worldIn, pos.offset(enumfacing));
+
                         worldIn.removeBlock(pos.offset(enumfacing));
                     }
                     else
@@ -265,12 +287,14 @@
             }
             else
             {
+                // TISCM break_bedrock stat
+                tryAddNearestPlayerStats(worldIn, pos.offset(enumfacing));
+
                 worldIn.removeBlock(pos.offset(enumfacing));
             }
 
             worldIn.playSound((EntityPlayer)null, pos, SoundEvents.BLOCK_PISTON_CONTRACT, SoundCategory.BLOCKS, 0.5F, worldIn.rand.nextFloat() * 0.15F + 0.6F);
         }
-
         return true;
     }
 
@@ -316,8 +340,14 @@
                 {
                     return false;
                 }
-
-                return !block.hasTileEntity();
+    
+                // [CM] Movable TE's CARPET-DeadlyMC
+                if(!block.hasTileEntity())
+                    return true;
+                else if(CarpetSettings.movableTileEntities)
+                    return isPushableTileEntityBlock(block);
+                else
+                    return false;
             }
             else
             {
@@ -329,6 +359,14 @@
             return false;
         }
     }
+    
+    // [CM] Movable TE's CARPET-DeadlyMC
+    private static boolean isPushableTileEntityBlock(Block block)
+    {
+        // Making MOVING_PISTON (BlockPistonMoving) pushable would not work as its createNewTileEntity()-method returns null
+        return block != Blocks.ENDER_CHEST && block != Blocks.ENCHANTING_TABLE && block != Blocks.END_GATEWAY
+                       && block != Blocks.END_PORTAL && block != Blocks.SPAWNER && block != Blocks.MOVING_PISTON;
+    }
 
     private boolean doMove(World worldIn, BlockPos pos, EnumFacing direction, boolean extending)
     {
@@ -371,6 +409,24 @@
                 --k;
                 aiblockstate[k] = iblockstate;
             }
+    
+            List<TileEntity> list1_TileEntities = Lists.<TileEntity>newArrayList();
+            // [CM] Movable TE's CARPET-DeadlyMC
+            if(CarpetSettings.movableTileEntities)
+            {
+                for (int i = 0; i < list.size(); ++i)
+                {
+                    BlockPos blockposnew = list.get(i);
+                    TileEntity tileentity = worldIn.getTileEntity(blockposnew);
+                    list1_TileEntities.add(tileentity);
+            
+                    if(tileentity != null)
+                    {
+                        worldIn.removeTileEntity(blockposnew);
+                        tileentity.markDirty();
+                    }
+                }
+            }
 
             for (int l = list.size() - 1; l >= 0; --l)
             {
@@ -379,7 +435,19 @@
                 blockpos3 = blockpos3.offset(enumfacing);
                 set.remove(blockpos3);
                 worldIn.setBlockState(blockpos3, Blocks.MOVING_PISTON.getDefaultState().with(FACING, direction), 68);
-                worldIn.setTileEntity(blockpos3, BlockPistonMoving.createTilePiston(list1.get(l), direction, extending, false));
+    
+                // [CM] Movable TE's CARPET-DeadlyMC
+                if (!CarpetSettings.movableTileEntities)
+                {
+                    worldIn.setTileEntity(blockpos3, BlockPistonMoving.createTilePiston(list1.get(l), direction, extending, false));
+                }
+                else
+                {
+                    TileEntity tilePiston = BlockPistonMoving.createTilePiston(list1.get(l), direction, extending, false);
+                    ((TileEntityPiston)tilePiston).setCarriedTileEntity(list1_TileEntities.get(l));
+                    worldIn.setTileEntity(blockpos3, tilePiston);
+                }
+                
                 --k;
                 aiblockstate[k] = iblockstate3;
             }
