--- a/net/minecraft/util/AxisRotation.java
+++ b/net/minecraft/util/AxisRotation.java
@@ -24,9 +24,23 @@
             return axis.getCoordinate(z, x, y);
         }
 
+        /**
+         * @reason Avoid expensive array/modulo operations
+         * @author JellySquid
+         */
         public EnumFacing.Axis rotate(EnumFacing.Axis axisIn)
         {
-            return AXES[Math.floorMod(axisIn.ordinal() + 1, 3)];
+            // return AXES[Math.floorMod(axisIn.ordinal() + 1, 3)];
+            switch (axisIn.ordinal()) {
+                case 0: //X
+                    return EnumFacing.Axis.Y;
+                case 1: //Y
+                    return EnumFacing.Axis.Z;
+                case 2: //Z
+                    return EnumFacing.Axis.X;
+            }
+
+            throw new IllegalArgumentException();
         }
 
         public AxisRotation reverse()
@@ -40,9 +54,23 @@
             return axis.getCoordinate(y, z, x);
         }
 
+        /**
+         * @reason Avoid expensive array/modulo operations
+         * @author JellySquid
+         */
         public EnumFacing.Axis rotate(EnumFacing.Axis axisIn)
         {
-            return AXES[Math.floorMod(axisIn.ordinal() - 1, 3)];
+            // return AXES[Math.floorMod(axisIn.ordinal() - 1, 3)];
+            switch (axisIn.ordinal()) {
+                case 0: //X
+                    return EnumFacing.Axis.Z;
+                case 1: //Y
+                    return EnumFacing.Axis.X;
+                case 2: //Z
+                    return EnumFacing.Axis.Y;
+            }
+
+            throw new IllegalArgumentException();
         }
 
         public AxisRotation reverse()
